% Created 2021-07-16 Fri 09:33
% Intended LaTeX compiler: pdflatex
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{grffile}
\usepackage{longtable}
\usepackage{wrapfig}
\usepackage{rotating}
\usepackage[normalem]{ulem}
\usepackage{amsmath}
\usepackage{textcomp}
\usepackage{amssymb}
\usepackage{capt-of}
\usepackage{hyperref}
\author{Aengus McGuinness, and Erika Pedersen}
\date{\today}
\title{}
\hypersetup{
 pdfauthor={Aengus McGuinness, and Erika Pedersen},
 pdftitle={},
 pdfkeywords={},
 pdfsubject={},
 pdfcreator={Emacs 26.3 (Org mode 9.1.9)}, 
 pdflang={English}}
\begin{document}

\tableofcontents

\section{Genome-visualizer.py: A Comprehensive Guide}
\label{sec:org65f9cf2}
\subsection{Visualizations}
\label{sec:orga7c17de}
Genome-visualizer.py is a python command line tool within the krepe package that
allows bioinformaticists to count the kmers with a .fasta, .fna, or .fastq
file and lets the user create bar graphs that represent all of the kmers 
at length k present in a file along with their abundance. It also provides
a simple uncompressed De Bruijn graph that is ideal for smaller values of
k and is a nice visualization of the genome.
\subsection{Execution}
\label{sec:org1dcf365}
After making the script executable using chmod +x you can run the
program with:
\$ ./\$PATH/Genome-visualizer.py (desired k-mer length) (file name) (bar graph) (de bruijn graph) (de-bruijn-meta-data)
Or using python3:
\$ python3 \$PATH/Genome-visualizer.py (desired k-mer length) (file name) (bar graph) (de bruijn graph) (de bruijn meta data)
\subsection{Tutorial}
\label{sec:org3ee5c2a}
Navigate to the tests folder and retrieve the Tuberculosis.fna file.
Genome-visualizer.py has two bar graph options: plot distribution, and
not_plot_distribution. The De Bruijn graph also has two arguements:
plot_de_bruijm and not_plot_de_bruijn. If the plot_de_bruijn arguement is used you must provide a meta data file that will provide the width, height, circle size, and stroke width for the de bruijn graph
An example usage would be:
\$ python3 Kmer\(_{\text{counter.py}}\) 25 tuberculosis.fna plot\(_{\text{distribution}}\) not\(_{\text{plot}}\)\(_{\text{de}}\)\(_{\text{bruijn}}\)
\end{document}